kernel/posix/pthread.c

```diff
@@ -0,0 +1,537 @@
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
+		return ESRCH;
+	}
+
+	pthread_mutex_lock(&thread->cancel_lock);
+	thread->cancel_pending = 1;
+
+	if (thread->cancel_state == PTHREAD_CANCEL_ENABLE) {
+		if (thread->state == PTHREAD_DETACHED) {
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+			pthread_mutex_lock(&thread->state_lock);
+			thread->state = PTHREAD_EXITED;
+			pthread_cond_broadcast(&thread->state_cond);
+			pthread_mutex_unlock(&thread->state_lock);
+		}
+
+		k_thread_abort((k_tid_t) thread);
+	} else {
+		pthread_mutex_unlock(&thread->cancel_lock);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Set thread scheduling policy and parameters.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setschedparam(pthread_t pthread, int policy,
+			  const struct sched_param *param)
+{
+	k_tid_t *thread = (k_tid_t *)pthread;
+	int new_prio, cur_prio;
+
+	if (thread == NULL) {
+		return ESRCH;
+	}
+
+	if (policy != SCHED_RR && policy != SCHED_FIFO) {
+		return EINVAL;
+	}
+
+	new_prio = posix_to_zephyr_priority(param->priority, policy);
+	cur_prio = k_thread_priority_get(*thread);
+
+	if (is_posix_prio_valid(new_prio, policy) == false ||
+	    new_prio == cur_prio) {
+		return EINVAL;
+	}
+
+	k_thread_priority_set(*thread, new_prio);
+
+	return 0;
+}
+
+/**
+ * @brief Initialise threads attribute object
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_init(pthread_attr_t *attr)
+{
+
+	if (attr->initialized == true) {
+		return EBUSY;
+	}
+
+	*attr = init_pthread_attrs;
+
+	return 0;
+}
+
+/**
+ * @brief Get thread scheduling policy and parameters
+ *
+ * See IEEE 1003.1
+ */
+int pthread_getschedparam(pthread_t pthread, int *policy,
+			  struct sched_param *param)
+{
+	k_tid_t thread = (k_tid_t)pthread;
+	u32_t  priority = k_thread_priority_get(thread);
+
+	param->priority = zephyr_to_posix_priority(priority, policy);
+	return 0;
+}
+
+/**
+ * @brief Terminate calling thread.
+ *
+ * See IEEE 1003.1
+ */
+void pthread_exit(void *retval)
+{
+	struct posix_thread *self = (struct posix_thread *)pthread_self();
+
+	/* Make a thread as cancelable before exiting */
+	pthread_mutex_lock(&self->cancel_lock);
+	if (self->cancel_state == PTHREAD_CANCEL_DISABLE) {
+		self->cancel_state = PTHREAD_CANCEL_ENABLE;
+		_thread_essential_clear();
+	}
+	pthread_mutex_unlock(&self->cancel_lock);
+
+	pthread_mutex_lock(&self->state_lock);
+	if (self->state == PTHREAD_JOINABLE) {
+		self->retval = retval;
+		self->state = PTHREAD_EXITED;
+		self->retval = retval;
+		pthread_cond_broadcast(&self->state_cond);
+	} else {
+		self->state = PTHREAD_TERMINATED;
+	}
+
+	pthread_mutex_unlock(&self->state_lock);
+	k_thread_abort((k_tid_t)self);
+}
+
+/**
+ * @brief Wait for a thread termination.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_join(pthread_t thread, void **status)
```

Comment:
For this function, it might be better to have a `int ret` that's set by each block comparing with `pthread->state`. This will reduce its size slightly, and have only one mutex_unlock/return block.

Reply:
sure i will modify this.